cudaarithm: async call to NPP fails

<!--
If you have a question rather than reporting a bug please go to http://answers.opencv.org where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

##### System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV => master 
- Operating System / Platform => Windows 10 Pro 64bit / CUDA 8.0
- Compiler => Visual Studio 2015 Win64
- Operating System / Platform => Ubuntu 64bit  / Jetson TX1 / CUDA 8.0
- Compiler => GCC 5.4.0

##### Detailed description
- test from ```opencv_test_cudaarithm``` fails
```
[==========] 10629 tests from 65 test cases ran. (34990 ms total)
[  PASSED  ] 10625 tests.
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] CUDA_Arithm/MeanStdDev.Async/0, where GetParam() = (NVIDIA Tegra X1, 128x128, whole matrix)
[  FAILED  ] CUDA_Arithm/MeanStdDev.Async/1, where GetParam() = (NVIDIA Tegra X1, 128x128, sub matrix)
[  FAILED  ] CUDA_Arithm/MeanStdDev.Async/2, where GetParam() = (NVIDIA Tegra X1, 113x113, whole matrix)
[  FAILED  ] CUDA_Arithm/MeanStdDev.Async/3, where GetParam() = (NVIDIA Tegra X1, 113x113, sub matrix)

 4 FAILED TESTS
```
- Also confirmed with Jetson TX2, Geforce GTX 650M, Geforce GTX 1060
- I cut out the minimum code from OpenCV and then the crash didn't reproduce.
- I googled up with the message "Microsoft C++ exception: NppStatus at memory location" which appeared on Visual Studio, and found [this post](https://devtalk.nvidia.com/default/topic/895161/gpu-accelerated-libraries/npp-row-and-column-filters/post/4746520/#4746520) from a nVidia person

> It seems that any time you issue a nppSetStream() call that in fact changes the underlying stream, you will need to issue a cudaDeviceSynchronize() call (first), before issuing the nppSetStream() call.

- This is exactly what is [happening inside OpenCV](https://github.com/opencv/opencv/blob/7e9578789b8f01cabf92d2c57691dbe65d6ee5df/modules/core/include/opencv2/core/private.cuda.hpp#L142)
- I added a macro to call ```cudaStreamSynchronize``` before ```nppSetStream``` and this solved the issue
- References
   - [Same situation still happens on CUDA 8.0](https://devtalk.nvidia.com/default/topic/999910/gpu-accelerated-libraries/npp-stream-crash/)
   - [Same explanation on NPP 9.1](http://docs.nvidia.com/cuda/npp/index.html)
> For this reason it is recommended that cudaDeviceSynchronize (or at least cudaStreamSynchronize) be called before making an nppSetStream call to change to a new stream ID. This will insure that any internal function calls that have not yet occurred will be completed using the current stream ID before it changes to a new ID.
   - [cudaStreamSynchronize](https://www.cs.cmu.edu/afs/cs/academic/class/15668-s11/www/cuda-doc/html/group__CUDART__STREAM_geb3b2f88b7c1cff8b67a998a3a41c179.html)

- I'll send a PR later
<!-- your description -->

##### Steps to reproduce
- run ```opencv_test_cudaarithm --gtest_filter=*MeanStdDev*Async*```
<!-- to add code example fence it with triple backticks and optional file extension
    ```.cpp
    // C++ code example
    ```
 or attach as .txt or .zip file
-->

I was also looking into this issue. In my case, using `cudaStreamSynchronize` nor `cudaDeviceSynchronize` solved the issue.

My testing environment is:
- Operating System / Platform => Windows 10 Pro 64bit / CUDA 8.0, 9.1
- Compiler => Visual Studio 2015
- GPU => GTX1080 with driver version 388.19

The error message on my machine is:
```
unknown file: error: C++ exception with description "OpenCV(3.4.1-dev) C:\Users\lee.namgoo\Repos\opencv_dev\opencv\modules\cudaarithm\src\reductions.cpp:162: error: (-217) Gpu API call: NPP_RANGE_ERROR [Code = -7] in function cv::cuda::meanStdDev" thrown in the test body.
```
The error message is identical before and after adding the synchronizing API.

--- a/modules/core/include/opencv2/core/private.cuda.hpp
+++ b/modules/core/include/opencv2/core/private.cuda.hpp
@@ -108,6 +108,8 @@ static inline void throw_no_cuda() { CV_Error(cv::Error::GpuNotSupported, "The l
 
 #else // HAVE_CUDA
 
+#define nppSafeSetStream(oldStream, newStream) { if(oldStream != newStream) { cudaStreamSynchronize(oldStream); nppSetStream(newStream); } }
+
 static inline void throw_no_cuda() { CV_Error(cv::Error::StsNotImplemented, "The called functionality is disabled for current build or platform"); }
 
 namespace cv { namespace cuda
@@ -139,13 +141,13 @@ namespace cv { namespace cuda
         inline explicit NppStreamHandler(Stream& newStream)
         {
             oldStream = nppGetStream();
-            nppSetStream(StreamAccessor::getStream(newStream));
+            nppSafeSetStream(oldStream, StreamAccessor::getStream(newStream));
         }
 
         inline explicit NppStreamHandler(cudaStream_t newStream)
         {
             oldStream = nppGetStream();
-            nppSetStream(newStream);
+            nppSafeSetStream(oldStream, newStream);
         }
 
         inline ~NppStreamHandler()

--- a/modules/cudaarithm/src/reductions.cpp
+++ b/modules/cudaarithm/src/reductions.cpp
@@ -157,8 +157,6 @@ void cv::cuda::meanStdDev(InputArray _src, OutputArray _dst, Stream& stream)
     BufferPool pool(stream);
     GpuMat buf = pool.getBuffer(1, bufSize, CV_8UC1);
 
-    NppStreamHandler h(StreamAccessor::getStream(stream));
-
     nppSafeCall( nppiMean_StdDev_8u_C1R(src.ptr<Npp8u>(), static_cast<int>(src.step), sz, buf.ptr<Npp8u>(), dst.ptr<Npp64f>(), dst.ptr<Npp64f>() + 1) );
 
     syncOutput(dst, _dst, stream);

--- a/modules/cudaarithm/src/reductions.cpp
+++ b/modules/cudaarithm/src/reductions.cpp
@@ -137,11 +137,12 @@ void cv::cuda::meanStdDev(InputArray _src, OutputArray _dst, Stream& stream)
     if (!deviceSupports(FEATURE_SET_COMPUTE_13))
         CV_Error(cv::Error::StsNotImplemented, "Not sufficient compute capebility");
 
-    const GpuMat src = getInputMat(_src, stream);
+    GpuMat src = getInputMat(_src, stream);
 
     CV_Assert( src.type() == CV_8UC1 );
 
-    GpuMat dst = getOutputMat(_dst, 1, 2, CV_64FC1, stream);
+    _dst.create(1, 2, CV_64FC1);
+    GpuMat dst = _dst.getGpuMat();
 
     NppiSize sz;
     sz.width  = src.cols;
@@ -156,6 +157,9 @@ void cv::cuda::meanStdDev(InputArray _src, OutputArray _dst, Stream& stream)
 
     BufferPool pool(stream);
     GpuMat buf = pool.getBuffer(1, bufSize, CV_8UC1);
+
+    // detail: https://github.com/opencv/opencv/issues/11063
+    //NppStreamHandler h(StreamAccessor::getStream(stream));
 
     nppSafeCall( nppiMean_StdDev_8u_C1R(src.ptr<Npp8u>(), static_cast<int>(src.step), sz, buf.ptr<Npp8u>(), dst.ptr<Npp64f>(), dst.ptr<Npp64f>() + 1) );
 

--- a/modules/cudaimgproc/src/histogram.cpp
+++ b/modules/cudaimgproc/src/histogram.cpp
@@ -107,7 +107,7 @@ namespace hist
 
 void cv::cuda::equalizeHist(InputArray _src, OutputArray _dst, Stream& _stream)
 {
-    GpuMat src = _src.getGpuMat();
+    GpuMat src = getInputMat(_src, _stream);
 
     CV_Assert( src.type() == CV_8UC1 );
 

--- a/modules/cudaimgproc/test/test_histogram.cpp
+++ b/modules/cudaimgproc/test/test_histogram.cpp
@@ -194,6 +194,23 @@ PARAM_TEST_CASE(EqualizeHist, cv::cuda::DeviceInfo, cv::Size)
     }
 };
 
+CUDA_TEST_P(EqualizeHist, Async)
+{
+    cv::Mat src = randomMat(size, CV_8UC1);
+
+    cv::cuda::Stream stream;
+
+    cv::cuda::GpuMat dst;
+    cv::cuda::equalizeHist(loadMat(src), dst, stream);
+
+    stream.waitForCompletion();
+
+    cv::Mat dst_gold;
+    cv::equalizeHist(src, dst_gold);
+
+    EXPECT_MAT_NEAR(dst_gold, dst, 3.0);
+}
+
 CUDA_TEST_P(EqualizeHist, Accuracy)
 {
     cv::Mat src = randomMat(size, CV_8UC1);

